### ISSUE TITLE
[Meta] Project refactoring

### ISSUE THREAD

**Issue by BoboTiG** [BoboTiG/ebook-reader-dict#458]
Note: the description is updated with comments and changes requested in comments.

The goal is to rework the `script` module to allow more flexibility and clearly separate concerns.

First, about the module name: `script`. It has been decided to change to `wikidict`.

## Overview

I would like to see the module splitted into 4 parts (each part will independent from others and can be replayed & extended easily).
This will also help leveraging multithreading to speed-up the whole process.

1. [x] Download the data (#466)
2. [x] Parse and store raw data (#469)
3. [x] Render templates and store results (#469)
4. [ ] Output to the proper eBook reader format

I have in mind a SQLite database where raw data will be stored and updated when needed.
Then, the parts will only use the data from the database. It should speed-up regenerating a whole dictionary when we update a template.

Then, each and every part will have its own CLI:

```shell
$ python -m wikidict --download ...
$ python -m wikidict --parse ...
$ python -m wikidict --render ...
$ python -m wikidict --output ...
```

And the all-in-one operation would be:
```
$ python -m wikidict --run ...
```

Side note: we could use an entry point to only having to type `wikidict` instead of `python -m wikidict`.

## Splitting get.py

Here we are talking about parts 1 and 2.

Part 1 is already almost fine as-is, we just need to move the code into its own submodule.
We could improve the CLI by allowing passing the Wiktionary dump date as argument, instead of relying on an envar.

Part 2 is only the mater of parsing the big XML file and storing raw data into a SQLite database. I am thinking of using this schema:

```
table: Word
fields:
    - word: varchar(256)
    - code: text
index on: word

table: Render
fields:
    - word_id: int
    - nature: varchar(16)
    - text: text
foreign key: word_id (Word._rowid_)
```

- The `Word` table will contain raw data from the Wiktionary.
- The `Render` table will be used to store the transformed text for a given word (after being cleaned up and where templates were processed). It will allow to have multiple texts for a given word (noun 1, noun 2, verb, adjective, ...).

We will have  one database per locale, located at `data/$LOCALE/$WIKIDUMP_DATE.db`.

At the download step, if no database exists, it will be retrieved from GitHub releases where they will be saved alongside dictionaries.
This is a cool thing IMO: everyone will have the good and up-to-date local database.
Of course, we will have options to skip it if the local file already exists or if we would like to force the download.

At the parse step, we will have to find a way to prevent parsing again if we run the command twice on the same Wiktionary dump.
I was thinking of using the [PRAGME user_version](https://www.sqlite.org/pragma.html#pragma_user_version) that would contain the Wiktionary dump date as integer.
It would be set only after the full parsing is done with success.

## Splitting convert.py

Here we are talking about parts 3 and 4.

Part 3 will call `clean()` and `process_templates()` on the `wikicode`. And store the result into the `rendered` field. This is the most time and CPU consuming part. It will be parallelized.

Part 4 will rethink how we are handling dictionary output to easily add more formats.

I was thinking of using a class with those methods (not really thought about it, I am just proposing the idea):

```python
class BaseFormat:

    __slots__ = {"locale", "output_dir"}

    def __init__(self, locale: str, output_dir: Path) -> None:
        self.locale = locale
        self.output_dir = output_dir
    
    def process(self) -> None:
        raise NotImplementedError()

    def save(self) -> None:
        raise NotImplementedError()


class KoboFormat(BaseFormat):
    def process(self, words) -> None:
        groups = self.make_groups(self.words)
        variants = self.make_variants(self.words)

        wordlist = []
        for word in words:
            wordlist.append(self.process_word(word))

        self.save(wordlist, groups, variants)

    def save(self, ...) -> None:
        ...
```

That part is way from being finished, but when we have a fully working format, in our code will will use that kind of code to generate the dict file:

```python
# Get all registered formats
formaters = get_formaters()

# Get all words from the database
words = get_words()

# And distribute the workload
from multiprocessing import Pool

def run(cls):
    formater = cls(locale, output_dir)
    formater.process(words)

with Pool(len(formaters)) as pool:
    pool.map(run_formatter, formaters))
```

**Comment by BoboTiG** [comment]
Everytinh may change, and there is missing details out there. But I open it now to get your feedbacks @lasconic.

**Comment by lasconic** [comment]
In general, I agree. 

In the details,
* `wikidict` sounds good 
* About the db, I would really like a way to have a word stored multiple times. I don't know how yet, but I would like to be able to separate the definitions of word by his "type" (noun, adj etc...). One way would be to store the word twice in the Kobo format, so maybe twice in the db as well.
* Could you elaborate on the notion of deleted words ?
* For a start, could we just link a db to a given wiktionary dump date ? and recreate db from scratch for a new dump date?
* I really like the idea to parallelize the clean/process template, part 3, because it's the longest right now and the most CPU intensive. But I believe you are refering to Part 4 with your multithreaded code ? It's nice if we have 10 formats in the future, but for now we already have thousands of words.

**Comment by BoboTiG** [comment]
1. ✔️ 
2. Good catch. It would make sense to use a separate table for rendered data. Like that, we could have multiple renders for a given word. But a word will be kept uniq as it is on the Wiktionary. WDYT?
3. Yesterday, the db was generated. Today, a new Wiktionary dump is available and the db will be updated with changes (noew words and modifications). But what about deleted words between those 2 dumps?
4. Yes we can. The file can the Wiktionary dump date + ".db".
5. The part 3 must be parallelized, and will be. Same for part 4, but at a lower level as the time is critical only on part 3.

**Comment by BoboTiG** [comment]
I updated the description with your suggestions. Do you agree on all technical names (DB fields, tables), step names, etc.?

**Comment by lasconic** [comment]
Steps ok.
process, run... I would go for run.
In  the first table, we agree that the section names are still in the wikicode right ?

**Comment by BoboTiG** [comment]
To start, we store the raw wikicode only. No processing is done.

Do you see a need to store sections? The code can be easily parsed when needed.

**Comment by lasconic** [comment]
No it's ok. I wanted to make sure that the information is still in, so we can split words in part 3 if needed.

**Comment by BoboTiG** [comment]
Out of curiosity, do you want to work on a part of the refactoring? Or I could handle it and ypu focus on the locale one?

**Comment by lasconic** [comment]
Ok. Let's do that. I will look into the locale one.
This one could be split in steps I guess.

**Comment by lasconic** [comment]
Is SQLite necessary ?

To separate parse and render: 
-- parse step could create a json file with wikicode `data_wikicode.json`, could be multithreaded
-- render would load `data_wikicode.json` and use multiple thread to render and store in the current data.json

What's the point of using a DB ? The main expected benefit is to run render/parse on single words for partial parse (for dump update/delete) or for partial render (if we change a template renderer) ? something else ?

**Comment by lasconic** [comment]
Can I help with something in particular for this issue? I will have less time in the coming weeks.

**Comment by BoboTiG** [comment]
You are right about the SQLite need, let's go with JSON.
I am light on free time, so I did not move foward yet. If you want to tackle a step, go ahead :)

**Comment by lasconic** [comment]
I did find another case for a database, parse and render can be done in parallel, parse filling a queue and render dequeuing it... but yes, JSON can be a first step.

**Comment by lasconic** [comment]
I had a quick look to --parse, assuming we just want to store word -> sections that we should consider. 
It means we need to parse the pages.xml file, then parse the code for each word in order to filter out unwanted sections by title, and then save to JSON. 
On my computer, just parsing the XML for french takes almost 6 minutes...
Parsing the XML + parsing the code with wikiparser takes 17 minutes (resulting json is 772MB), so 11 minutes for the wikiparser

The current get.py 
```
python -Wd -m scripts --locale fr --fetch-only
```
takes 25 minutes, so 8 minutes for our own template rendering.

The wikiparser takes quite some time and we will parse the wikicode twice, once in parse (the whole wikicode for section detection) and once in convert (only the interesting sections). It will slow down the whole process for sure... Maybe we should try to get rid of wikiparser in the parse step.

**Comment by lasconic** [comment]
I took a different approach. --parse will just read the XML and remove words that do not contain `headsections` (maybe "sections" ?) with a pure string comparison, no wikitextparser involved. Then it stores to a json file word->wikicode. It takes a bit less than 10 minutes
 ```
>>> Saved 1,823,776 words into data/fr/data_wikicode-20201201.json (966M)
main(): 0:09:50.835394
```

Then --render will do the rest of the work. And we are down to the same time.
>>> Saved 1,539,741 words into data/fr/data-20201201.json
>>> Render done!
main(): 0:15:40.053682

**Comment by BoboTiG** [comment]
Parsing that takes 10 minutes seems a bit too high. My last benchmark (just parsing the XML) was taking 2 minutes.
Do you know what part of the parsing is the bottleneck?

The render could be parallelized ad hopefully we could get better numbers. We will have to do some testing for that.

**Comment by lasconic** [comment]
> Parsing that takes 10 minutes seems a bit too high. My last benchmark (just parsing the XML) was taking 2 minutes.
> Do you know what part of the parsing is the bottleneck?
My CPU maybe... (mac 2,3 GHz Intel Core i7)

**Comment by lasconic** [comment]
It was faster on the second run with the profiler running... Less than 6 minutes
https://gist.github.com/lasconic/063a2e71a4300c2815251d68832270d8

**Comment by BoboTiG** [comment]
About the final step: converting to the Kobo dictionary.

You talked about {{.df}} files in #409. I was wondering if, for that ticket only, we just implement the code that outputs to such format. Doing so would let pyglossary to do all the work then (Kobo, Kindle, ...). So the last step would just the amtter of outputing to the good {{.df}} format.

**Comment by BoboTiG** [comment]
Well, we are using a custom HTML code for the Kobo, we will need to test the pyglossary output before doing such move.

**Comment by lasconic** [comment]
I would keep  the Kobo output as it and add df. I'm not sure if pyglossary can be called from python or should be called from the CI.

Render with 7 threads is twice faster. PR to come.

**Comment by lasconic** [comment]
Mostly done. Only testing is messing... How do you want to tackle it ?
Big project starting tomorrow or later this week. The last PR could be my last push for a while.

**Comment by BoboTiG** [comment]
Hmm `test_N_*.py` files will be a pain, I will handle them. If you can migrate `test_$LOCALE.py` it would be great. But again, if you do not have time, I will have some on my side, so not a big deal.

If you want to tackle another issue, go ahead too :)

**Comment by lasconic** [comment]
`test_$LOCALE.py` was easy ;) see PR #478

**Comment by BoboTiG** [comment]
I can close the issue now, thanks a lot for your help, it was awesome 💪

**Comment by BoboTiG** [comment]
The refactoring is finished 🍾 
Tests coverage is at 100% (module `arabiser.py`). I think we should rework release descriptions to include all dictionaries (when available) or simply list donwload files.
